src/ws/session-utils.c

```diff
@@ -984,3 +938,67 @@ fd_remap (const int *remap_fds,
   if (fdwalk (closefd, &n_remap_fds) < 0)
     abort_with_message ("couldn't close all file descriptors");
 }
+
+int
+spawn_and_wait (const char **argv, const char **envp,
+                const int *remap_fds, int n_remap_fds,
+                uid_t uid, gid_t gid)
+{
+  pid_t child;
+
+  child = fork ();
+  if (child == -1)
+    abort_with_message ("cockpit-session: fork() failed: %m");
+
+  if (child == 0)
+    {
+      /* This is the child process.  Do preparation, and exec(). */
+      if (setresgid (gid, gid, gid) != 0)
+        abort_with_message ("setresgid: couldn't set gid to %u: %m\n", (int) gid);
+
+      if (setresuid (uid, uid, uid) != 0)
+        abort_with_message ("setresgid: couldn't set uid to %u: %m\n", (int) gid);
```

Comment:
Again, this commit replaces the previously safe approach with a dangerous one -- setuid() is terminal, while setresuid() is reversible. At which point `CAP_SETUID` is dropped, so that the user session can't get back to root?

This commit increases the code size and makes it much harder to audit, so please add a good justification why this happens.